README
Net-LDAPxs version 1.20
=======================

Net::LDAPxs is using XS code to glue LDAP C API Perl code. The purpose of developing this module is to thoroughly improve the performance of Net::LDAP.  According to the simple test using L<Devel::NYTProf>, it can enhance the performance by nearly 30%.

In order to benefit the migration from Net::LDAP to Net::LDAPxs, functions and user interfaces of Net::LDAPxs keep the same as Net::LDAP, which means people who migrate from Net::LDAP to Net::LDAPxs are able to leave their code unchanged excepting altering the module name. 

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Copyright (C) 2008-2010 by Pan Yu. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.



Changes
Revision history for Perl extension Net::LDAPxs.

0.01 Thu Sep 24 13:46:29 2009
	- original version; created by h2xs 1.23 with options
		-O -n Net::LDAPxs ./Net-LDAPxs/xslib/ldapxs.h
0.02 Sat Nov 28 2009
	- add new functions
	- fix bugs
	- update document
	- improve error detection
0.03 Sun Nov 29 2009
	- add function delete
	- fix bugs
	- update document
0.04 Wed Dec 21 2009
	- add control function
		server side sorting countrol
	- fix bugs
	- update document
0.05 Wed Dec 30 2009
	- modify function
	- moddn function
1.20 Tue Jul 13 2010
	- fundamentally change the underlying mechanism
	- add a new module Net::LDAPxs::Exception
	- improved the typemap
	- improved exceptions handling system
	- fixed a number of bugs



lib/Net/LDAPxs/Entry.pm
package Net::LDAPxs::Entry;

use 5.006;
use strict;
use vars qw($VERSION);

$VERSION = '0.01';


sub dn { shift->{objectName}; }

sub attributes {
	my $self = shift;

	map { $_->{type} } @{$self->{attributes}};
}

sub get_value {
	my $self = shift;
	my $type = shift;

	my $attrs = _build_attrs($self);
	@{$attrs->{$type}};
}

sub _build_attrs {
	+{ map { $_->{type}, $_->{vals} }  @{$_[0]->{attributes}} };
}


1;

__END__


=head1 NAME

Net::LDAPxs::Entry - An LDAP entry object

=head1 SYNOPSIS

  use Net::LDAPxs;
  
  $ldap = Net::LDAPxs->new ( $host );
  $msg = $ldap->search ( @search_args );
  
  @entries = $msg->entries();

  foreach my $entry (@entries) {
      foreach my $attr ($entry->attributes()) {
          foreach my $val ($entry->get_value($attr)) {
              print "$attr, $val\n";
          }
      }
  }

=head1 DESCRIPTION

The B<Net::LDAPxs::Entry> object represents a single entry in the
directory.  It is a container for attribute-value pairs.

=head1 METHODS

=over 4

=item attributes ( OPTIONS )

Return a list of attributes in this entry

=item dn ( )

Get the DN of the entry.

=item get_value ( ATTR )

Get the values for the attribute C<ATTR>. In a list context returns
all values for the given attribute, or the empty list if the attribute
does not exist. In a scalar context returns the first value for the
attribute.

=back

=head1 ACKNOWLEDGEMENTS

This document is based on the document of L<Net::LDAP::Entry>

=head1 AUTHOR

Pan Yu <xiaocong@vip.163.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2009 by Pan Yu. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Net/LDAPxs/Search.pm
package Net::LDAPxs::Search;

use 5.006;
use strict;
use vars qw($VERSION);
use Net::LDAPxs::Entry;

$VERSION = '0.01';


sub count { $#{shift->{entries}}+1; }

sub entries {
	my $self = shift;

	($self->count > 0) ? @{$self->{entries}} : undef;
}

sub shift_entry {
	my $self = shift;

	($self->count > 0) ? shift @{$self->{entries}} : undef;
}

sub pop_entry {
	my $self = shift;

	($self->count > 0) ? pop @{$self->{entries}} : undef;
}

sub all_entries { goto &entries }

sub code { }

1;

__END__

=head1 NAME

Net::LDAPxs::Search - Object returned by Net::LDAPxs search method

=head1 SYNOPSIS

  use Net::LDAPxs;
  
  $msg = $ldap->search( @search_args );
  
  @entries = $msg->entries;

=head1 DESCRIPTION

A B<Net::LDAPxs::Search> object is returned from the
L<search|Net::LDAPxs/item_search> method of a L<Net::LDAPxs> object. It is
a container object which holds the results of the search.

=head1 METHODS

=over 4

=item count

Returns the number of entries returned by the server.

=item entries

Return an array of L<Net::LDAPxs::Entry> objects that were returned from
the server.

=item pop_entry

Pop an entry from the internal list of L<Net::LDAPxs::Entry> objects for
this search. If there are no more entries then C<undef> is returned.

This call will block if the list is empty, until the server returns
another entry.

=item shift_entry

Shift an entry from the internal list of L<Net::LDAPxs::Entry> objects
for this search. 

This call will block if the list is empty, until the server returns
another entry.

=back

=head1 ACKNOWLEDGEMENTS

This document is based on the document of L<Net::LDAP::Search>

=head1 AUTHOR

Pan Yu <xiaocong@vip.163.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2009 by Pan Yu. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Net/LDAPxs/Control.pm
#
# Copyright (c) 2008-2009 Pan Yu (xiaocong@vip.163.com). 
# All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
package Net::LDAPxs::Control;

use 5.006;
use strict;
use vars qw($VERSION);
use vars qw( $DEFAULT_CONTROL_CRITICAL
			 $LDAP_CONTROL_SORTREQUEST
		);

$VERSION = '0.01';

$DEFAULT_CONTROL_CRITICAL = 0;
$LDAP_CONTROL_SORTREQUEST = '1.2.840.113556.1.4.473';

my $error = {
		'die'	=> sub { require Carp; Carp::croak(@_); },
		'warn'	=> sub { require Carp; Carp::carp(@_); }
};

my %controls = (
		$LDAP_CONTROL_SORTREQUEST => 1
);

sub _error {
	my $error_code = shift;
	my $error_msg = shift;

	$error->{$error_code}($error_msg);
}

sub _check_options {
	my $arg_ref = shift;

	if (grep { /^-/ } keys %$arg_ref) {
		_error('die', "Leading - for options is NOT supported");
	}
	if (exists $arg_ref->{'type'}) {
		if (exists $controls{$arg_ref->{type}}) {
			$arg_ref->{'type'} = $controls{$arg_ref->{type}};
		}else{
			_error('die', "The control type $arg_ref->{type} is currently not supported");	
		}
	}else{
		_error('die', "Control type option is required");
	}
	if (!exists $arg_ref->{'value'}) {
		_error('die', "Control value option is required");
	}
	$arg_ref->{'critical'} = $arg_ref->{'critical'} || $DEFAULT_CONTROL_CRITICAL;
}

sub new {
	my $class = shift;
	my $type = ref($class) || $class;
	my $arg_ref = { @_ };

	_check_options($arg_ref);

	$arg_ref;
}


1;

__END__


=head1 NAME

Net::LDAPxs::Control - LDAPv3 control extension

=head1 SYNOPSIS

  use Net::LDAPxs::Control;

  $ctrl = Net::LDAPxs::Control->new(
          type  => '1.2.840.113556.1.4.473',
          value => 'sn -cn',
          critical => 0
          );

  $msg = $ldap->search( base    => $base,
                        control => $ctrl );
  
=head1 DESCRIPTION

The C<Net::LDAPxs::Control> is for LDAPv3 control extension.

=head1 CONSTRUCTORS

=over 4

=item new ( ARGS )

ARGS is a list of name/value pairs, valid arguments are:

=item type

A dotted-decimal representation of an OBJECT IDENTIFIER which
uniquely identifies the control. This prevents conflicts between
control names.

=item value

Optional information associated with the control. It's format is specific
to the particular control.

=item critical

A boolean value, if TRUE and the control is unrecognized by the server or
is inappropriate for the requested operation then the server will return
an error and the operation will not be performed.

If FALSE and the control is unrecognized by the server or
is inappropriate for the requested operation then the server will ignore
the control and perform the requested operation as if the control was
not given.

If absent, FALSE is assumed.

=back

=head1 ACKNOWLEDGEMENTS

This document is based on the document of L<Net::LDAP::Control>

=head1 AUTHOR

Pan Yu <xiaocong@vip.163.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2009 by Pan Yu. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Net/LDAPxs/Exception.pm
package Net::LDAPxs::Exception;

use 5.006;
use strict;
use vars qw($VERSION);

$VERSION = '1.00';


sub code {
	return 1 if shift->{code} != 0;
}

sub errstr { 
	my $self = shift;

	"Error code($self->{code}): $self->{mesg}";
}

1;

__END__

=head1 NAME

Net::LDAPxs::Exception - Object handling the exceptions

=head1 SYNOPSIS

  use Net::LDAPxs;
  
  $mesg = $ldap->search( $dn, password => "secret" );
  die $mesg->errstr if $mesg->code;
  
=head1 DESCRIPTION

When exception happens, a B<Net::LDAPxs::Exception> object is returned 
from the methods of a L<Net::LDAPxs> object. It is used for displaying 
both error codes and error messages sent from LDAP server.

=head1 METHODS

=over 4

=item code

This method is for detecting whether there is an exception.

=item errstr

Method C<errstr> is in charge of displaying the error messages sent 
from the LDAP server.

=back

=head1 AUTHOR

Pan Yu <xiaocong@vip.163.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2010 by Pan Yu. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Net/LDAPxs.pm
#
# Copyright (c) 2008-2010 Pan Yu (xiaocong@vip.163.com). 
# All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#

package Net::LDAPxs;

use strict;

use Exporter;
use DynaLoader;
use vars qw($VERSION);
use vars qw($DEFAULT_LDAP_VERSION $DEFAULT_LDAP_PORT $DEFAULT_LDAP_SCHEME);
use Net::LDAPxs::Exception;

$VERSION = '1.20';

our @ISA = qw(Exporter DynaLoader);

our @EXPORT = ( );
our @EXPORT_OK = qw(
	new bind unbind search add compare delete modify moddn
);

bootstrap Net::LDAPxs;

$DEFAULT_LDAP_VERSION	= 3; 
$DEFAULT_LDAP_PORT		= 389; 
$DEFAULT_LDAP_SCHEME	= 'ldap'; 


my $error = {
		'die'	=> sub { require Carp; Carp::croak(@_); },
		'warn'	=> sub { require Carp; Carp::carp(@_); }
};

sub _error {
	my $error_code = shift;
	my $error_msg = shift;

	$error->{$error_code}($error_msg);
}

sub _check_options {
	my $arg_ref = shift;

	if (grep { /^-/ } keys %$arg_ref) {
		_error('die', "Leading - for options is NOT supported");
	}
	$arg_ref->{'port'} ||= $DEFAULT_LDAP_PORT;
	$arg_ref->{'version'} ||= $DEFAULT_LDAP_VERSION;
	$arg_ref->{'scheme'} ||= $DEFAULT_LDAP_SCHEME;
}

sub new {
	my $class = shift;
	my $host = shift;
	my $arg_ref = { @_ };

	_check_options($arg_ref);

	my $port = $arg_ref->{'port'};
	my $version = $arg_ref->{'version'};
	my $scheme = $arg_ref->{'scheme'};
	$arg_ref->{'host'} = $host;

	return _new($class, $arg_ref);
}

my %async = (0 => 0, 1 => 1);
sub bind {
	my $self = shift;
	my $binddn = shift;
	my $arg_ref = { @_ };

	my $opt;
	$opt->{binddn} = $binddn;
	$opt->{bindpw} = $arg_ref->{password} if defined $arg_ref->{password};
	$opt->{async} = $async{$arg_ref->{async} || 0};

	$self->_bind($opt);
}

sub unbind {
	shift->_unbind();
}

my %scope = ('base' => 0, 'one' => 1, 'sub' => 2, 'children' => 3);

sub search {
	my $self = shift;
	my $arg_ref = { @_ };

	my $opt;
	require Net::LDAPxs::Search;
	$opt->{base} = $arg_ref->{base};
	$opt->{filter} = $arg_ref->{filter};
	$opt->{async} = $async{$arg_ref->{async}} || 0;
	
	if (exists $arg_ref->{scope}) {
		my $scope = lc $arg_ref->{scope};
		$opt->{scope} = (exists $scope{$scope}) ? $scope{$scope} : 2;
	}else{
		$opt->{scope} = 2;
	}
	$opt->{sizelimit} ||= 0;
	$opt->{attrs} = $arg_ref->{attrs} if (defined $arg_ref->{attrs});
	# process control if present
	$opt->{control} = $arg_ref->{control} if (defined $arg_ref->{control}); 
	my $mesg = $self->_search($opt);

	# process callback if present
	if (defined $arg_ref->{callback}) {
		foreach my $entry ($mesg->entries) {
			$arg_ref->{callback}($mesg, $entry);
		}
	}
	return $mesg;
}

sub add {
	my $self = shift;
	my $dn = shift;
	my $arg_ref = { @_ };

	if (exists $arg_ref->{attrs}) {
		$self->_add($dn, $arg_ref->{attrs});
	}else{
		_error('die', "Option 'attrs' is required when using 'add' function");
	}
}

my %operation = ( 'add' => 0, 'delete' => 1, 'replace' => 2, 'increment' => 3 );
sub modify {
	my $self = shift;
	my $dn = shift;
	my $arg_ref = { @_ };
	my @options;

	# in order to be compatible with Net::LDAP
	$arg_ref = $arg_ref->{changes} if (exists $arg_ref->{changes});

	foreach my $op (qw(add delete replace increment)) {
		next unless exists $arg_ref->{$op};
		my ($key, $val);
		while (($key, $val) = each %{$arg_ref->{$op}}) {
			my $attrs;
			$attrs->{'changetype'} = $operation{$op};
			$attrs->{'type'} = $key;
			if (ref($val) eq 'ARRAY') {
				$attrs->{'vals'} = $val;
			}else{
				$attrs->{'vals'} = [ $val ];
			}
			push (@options, $attrs);
		}
	}
	$self->_modify($dn, \@options);
}

sub moddn {
	my $self = shift;
	my $dn = shift;
	my $arg_ref = { @_ };

	_error('die', "Option 'newrdn' is required") 
		unless exists $arg_ref->{newrdn};
	$arg_ref->{deleteoldrdn} ||= 1;

	$self->_moddn($dn, $arg_ref);
}

sub compare {
	my $self = shift;
	my $dn = shift;
	my $arg_ref = { @_ };

	$self->_compare($dn, $arg_ref->{attr}, $arg_ref->{value});
}

sub delete {
	my $self = shift;
	my $dn = shift;

	$self->_delete($dn);
}


1;

__END__

=head1 NAME

Net::LDAPxs - XS version of Net::LDAP

=head1 SYNOPSIS

  use Net::LDAPxs;

  $ldap = Net::LDAPxs->new('www.example.com');

  $ldap->bind('cn=Manager,dc=example,dc=com', password => 'secret');

  $mesg = $ldap->search( base   => 'ou=language,dc=example,dc=com',
                         filter => '(|(cn=aperture)(cn=shutter_speed))'
                       );

  @entries = $mesg->entries();

  foreach my $entry (@entries) {
      foreach my $attr ($entry->attributes()) {
          foreach my $val ($entry->get_value($attr)) {
              print "$attr, $val\n";
          }
      }
  }

  $ldap->unbind;

=head1 DESCRIPTION

Net::LDAPxs is using XS code to glue LDAP C API Perl code. The purpose of 
developing this module is to thoroughly improve the performance of Net::LDAP. 
According to the simple test using L<Devel::NYTProf>, it can enhance the 
performance by nearly 30 times.
In order to benefit the migration from Net::LDAP to Net::LDAPxs, functions and 
user interfaces of Net::LDAPxs keep the same as Net::LDAP, which means people 
who migrate from Net::LDAP to Net::LDAPxs are able to leave their code 
unchanged excepting altering the module name.

=head1 CONSTRUCTOR

=item new ( HOST, OPTIONS )

HOST can be a host name or an IP address without path information.

=over 4

=item port => ( number ) (Default: 389)

Port connect to the LDAP server.

=item scheme => 'ldap' | 'ldaps' | 'ldapi' (Default: ldap)

=back

B<Example>

  $ldap = Net::LDAPxs->new('www.example.com',
                           port    => '389',
                           scheme  => 'ldap',
                           version => 3
                          );

=head1 METHODS

Currently, not all methods of Net::LDAP are supported by Net::LDAPxs.
Here is a list of implemented methods.

=item bind ( DN, OPTIONS )

=over 4

=item async => 1

Perform the bind operation asynchronously.

B<Example>

  $mesg = $ldap->bind('cn=Manager,dc=example,dc=com', password => 'secret');
  die $mesg->errstr if $mesg->code;

=item unbind ( )

=back

B<Example>

  $ldap->unbind;

=item search ( ID, OPTIONS )

=over 4

=item base => ( DN )

A base option is a DN which is the start search point.

=item filter => ( a string )

A filter is a string which format complies the RFC1960.

=back

B<Example>

  (cn=Babs Jensen)
  (!(cn=Tim Howes))
  (&(objectClass=Person)(|(sn=Jensen)(cn=Babs J*)))
  (o=univ*of*mich*)

=over 4

=item scope => 'base' | 'one' | 'sub'

The default value is 'sub' which means it will search all subtrees. 'base' means only 
search the base object. 'one' means only search one level below the base object.

=item sizelimit => ( number )

A sizelimit is the maximum number of entries will be returned as a result of the 
search. The default value is 0, denots no restriction is applied.

=item attrs => ( attributes )

A list of attributes to be returned for each entry. The value is normally a reference 
to an array which contains the preferred attributes.

=back

B<Example>

  $mesg = $ldap->search( base      => 'ou=language,dc=example,dc=com',
                        filter    => '(|(cn=aperture)(cn=shutter_speed))',
                        scope     => 'one',
                        sizelimit => 0,
                        attrs     => \@attrs
                      );
  die $mesg->errstr if $mesg->code;

=over 4

=item control => ( CONTROL )

A control is a reference to a HASH which may contain the three elements "type", "value" and "critical". 

For more information see L<Net::LDAPxs::Control>.

=back

B<Example>

  use Net::LDAPxs::Control;

  $ctrl = Net::LDAPxs::Control->new(
          type  => '1.2.840.113556.1.4.473',
          value => 'sn -cn',
          critical => 0
          );

  $msg = $ldap->search( base    => $base,
                        control => $ctrl );

=item compare ( DN, OPTIONS )

Compare values in an attribute in the entry given by DN on the server. DN is a string.
If the compare is failed, errstr() method can be used to fetch the reason for the failure.

=over 4

=item attrs => attributeType

The name of the attribute type to compare.

=item value => attributeValue

The attribute value to compare with.

=back

B<Example>

  $mesg = $ldap->compare( 'ou=people,dc=example,dc=com',
                          attr  => 'gidNumber',
                          value => '65534'
                        );
  die $mesg->errstr if $mesg->code;

=item add ( DN, OPTIONS )

Add a new entry to the LDAP directiory. DN is a string.

=over 4

=item attrs => VALUE

C<VALUE> should be a hash reference.

=back

B<Example>

  my %attrs = (
    uid => 'Lionel',
    cn  => 'Lionel',
    sn  => 'Luthor',
    uidNumber    => '65534',
    gidNumber    => '65534',
    homeDirectory => '/home/Lionel',
    loginShell  => '/bin/bash',
    objectClass => [qw(inetOrgPerson posixAccount top)]
  );

  $mesg = $ldap->add( 'uid=Lionel,ou=people,dc=example,dc=com',
                      attrs => \%attrs );
  die $mesg->errstr if $mesg->code;

=item delete ( DN )

Delete the entry given by C<DN> from the server. C<DN> is a string.

B<Example>

  $mesg = $ldap->delete( 'uid=Lionel,ou=people,dc=example,dc=com' );
  die $mesg->errstr if $mesg->code;

=item moddn ( DN, OPTIONS )

Rename the entry given by C<DN> which should be a string.

=over 4

=item newrdn => RDN

This value should be a new RDN to assign to C<DN>.

=item deleteoldrdn => 1

This option should be passed if the existing RDN is to be deleted.

=item newsuperior => NEWDN

If given this value should be the DN of the new superior for C<DN>.

=back

B<Example>

  $mesg = $ldap->moddn( uid=Lionel,ou=people,dc=example,dc=com, 
                        newrdn => 'uid=Peter' );
  die $mesg->errstr if $mesg->code;

=item modify ( DN, OPTIONS )

Modify the contents of the entry given by C<DN> which should be a string.

=over 4

=item add => { ATTR => VALUE, ... }

Add more attributes or values to the entry. C<VALUE> should be a
string if only a single value is wanted in the attribute, or a
reference to an array of strings if multiple values are wanted.

  %attrs = ( cn => ['buy', 'purchase'],
		     description => 'to own something' );
  $mesg = $ldap->modify( $dn, add => \%attrs );
  die $mesg->errstr if $mesg->code;

=item delete => [ ATTR, ... ]

Delete complete attributes from the entry.

  # Delete attributes
  $mesg = $ldap->modify( $dn, delete => ['member','description'] );
  die $mesg->errstr if $mesg->code;
  
  # Delete a group of attributes
  %attrs = (
      cn => ['Lex', 'Lionel'],
      sn => 'Luther'
  );
  $mesg = $ldap->modify( $dn, delete => \%attrs );

=item delete => { ATTR => VALUE, ... }

Delete individual values from an attribute. C<VALUE> should be a
string if only a single value is being deleted from the attribute, or
a reference to an array of strings if multiple values are being
deleted.

If C<VALUE> is a reference to an empty array or all existing values
of the attribute are being deleted, then the attribute will be
deleted from the entry.

  $mesg = $ldap->modify( $dn,
    delete => {
      description => 'List of members',
      member      => [
        'cn=member1,ou=people,dc=example,dc=com',    # Remove members
        'cn=member2,ou=people,dc=example,dc=com',
      ],
      seeAlso => [],   # Remove attribute
    }
  );

=item replace => { ATTR => VALUE, ... }

Replace any existing values in each given attribute with
C<VALUE>. C<VALUE> should be a string if only a single value is wanted
in the attribute, or a reference to an array of strings if multiple
values are wanted. A reference to an empty array will remove the
entire attribute. If the attribute does not already exist in the
entry, it will be created.

  $mesg = $ldap->modify( $dn,
    replace => {
      description => 'New List of members', # Change the description
      member      => [ # Replace whole list with these
        'cn=member1,ou=people,dc=example,dc=com',   
        'cn=member2,ou=people,dc=example,dc=com',
      ],
      seeAlso => [],   # Remove attribute
    }
  );

=item increment => { ATTR => VALUE, ... }

Atomically increment the existing value in each given attribute by the
provided C<VALUE>. The attributes need to have integer syntax, or be
otherwise "incrementable". Note this will only work if the server
advertizes support for LDAP_FEATURE_MODIFY_INCREMENT. 

  $mesg = $ldap->modify( $dn,
    increment => {
      uidNumber => 1 # increase the uidNumber by 1
    }
  );

=item changes => [ OP => [ ATTR => VALUE ], ... ]

This is an alternative to B<add>, B<delete>, B<replace> and B<increment>
where the whole operation can be given in a single argument. C<OP>
should be B<add>, B<delete>, B<replace> or B<increment>. C<VALUE> should
be either a string or a reference to an array of strings, as before.

Use this form if you want to control the order in which the operations
will be performed.

  $mesg = $ldap->modify( $dn,
    changes => [
      add => [
        description => 'A description',
        member      => $newMember,
      ],
      delete => [
        seeAlso => [],
      ],
      add => [
        anotherAttribute => $value,
      ],
    ]
  );

=head1 DEVELOPMENT STAGE

This module is currently in production. The advanced features will be available soon.

=head1 BUGS and RECOMMENDATIONS

Any bugs and recommendation is welcome. Please send directly to my email address 
listed below. Bugs and functions will be updated at least every one month. 

=head1 ACKNOWLEDGEMENTS

A special thanks to Larry Wall <larry@wall.org> for convincing me that no 
development could be made to the Perl community without people's contribution.

=head1 AUTHOR

Pan Yu <xiaocong@vip.163.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2010 by Pan Yu. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

